pyanaconda/ui/tui/spokes/time_spoke.py

```diff
@@ -19,34 +19,113 @@
 # Red Hat Author(s): Martin Sivak <msivak@redhat.com>
 #
 
+import logging
+log = logging.getLogger("anaconda")
 from pyanaconda.ui.categories.localization import LocalizationCategory
-from pyanaconda.ui.tui.spokes import NormalTUISpoke
+from pyanaconda.ui.tui.spokes import NormalTUISpoke, EditTUIDialog
 from pyanaconda.ui.tui.simpleline import TextWidget, ColumnWidget
 from pyanaconda.ui.common import FirstbootSpokeMixIn
 from pyanaconda import timezone
+from pyanaconda import ntp
+from pyanaconda import constants
 from pyanaconda.i18n import N_, _, C_
 from pyanaconda.constants_text import INPUT_PROCESSED
+from pyanaconda.threads import threadMgr, AnacondaThread
+from pyanaconda.flags import flags
 
-class TimeZoneSpoke(FirstbootSpokeMixIn, NormalTUISpoke):
-    title = N_("Timezone settings")
+from collections import OrderedDict
+
+NTP_SERVER_STATES = {
+    constants.NTP_SERVER_OK: _("status: working"),
+    constants.NTP_SERVER_NOK: _("status: not working"),
+    constants.NTP_SERVER_QUERY: _("checking status")
+}
+
+def format_ntp_status_list(servers):
+    status_list = []
+    for server, server_state in servers.items():
+        status_list.append("%s (%s)" % (server, NTP_SERVER_STATES[server_state]))
+    return status_list
+
+__all__ = ["TimeSpoke"]
+
+class TimeSpoke(FirstbootSpokeMixIn, NormalTUISpoke):
+    title = N_("Time settings")
     category = LocalizationCategory
 
     def __init__(self, app, data, storage, payload, instclass):
         NormalTUISpoke.__init__(self, app, data, storage, payload, instclass)
+        self._timezone_spoke = None
+        # we use an ordered dict to keep the NTP server insertion order
+        self._ntp_servers = OrderedDict()
```

Comment:
That's a good point - it would not be an issue if only adding was supported (as the thread only run on any initial servers and then on any server added), but since we also support removing of NTP servers, the thread might indeed try to update status of a server that is no longer in the dictionary, resulting in a traceback.